controller/resource_test.go

```diff
@@ -595,3 +750,494 @@ func TestParseServiceNameFromTaskDefinitionARN(t *testing.T) {
+
+			for _, r := range resources {
+				s := r.(*ServiceInfo)
+
+				// use a unique secret for each service
+				accessorID, err := uuid.GenerateUUID()
+				require.NoError(t, err)
+				secretID, err := uuid.GenerateUUID()
+				require.NoError(t, err)
+				s.SecretsManagerClient = &mocks.SMClient{Secret: &secretsmanager.GetSecretValueOutput{
+					Name:         aws.String(s.ServiceName.Name),
+					SecretString: aws.String(fmt.Sprintf(`{"accessor_id":"%s","token":"%s"}`, accessorID, secretID)),
+				}}
+
+				// call Reconcile() to create policies and tokens for each service
+				require.NoError(t, s.Reconcile())
+			}
+
+			// call List() to get updated cluster state
+			resources, err = lister.List()
+			require.NoError(t, err)
+
+			// inspect the state of each service reported by the controller
+			obsServices := make([]*ServiceInfo, 0, len(resources))
+			for _, r := range resources {
+				s := r.(*ServiceInfo)
+
+				// set only the expected ACL fields in the service state
+				var policies []*api.ACLPolicyListEntry
+				for _, policy := range s.ServiceState.ACLPolicies {
+					policies = append(policies, &api.ACLPolicyListEntry{
+						Partition:   policy.Partition,
+						Namespace:   policy.Namespace,
+						Name:        policy.Name,
+						Description: policy.Description,
+					})
+				}
+				s.ServiceState.ACLPolicies = policies
+
+				var tokens []*api.ACLTokenListEntry
+				for _, token := range s.ServiceState.ACLTokens {
+					tokens = append(tokens, &api.ACLTokenListEntry{
+						Partition:   token.Partition,
+						Namespace:   token.Namespace,
+						Description: token.Description,
+					})
+				}
+				s.ServiceState.ACLTokens = tokens
+
+				obsServices = append(obsServices, s)
+			}
+
+			// check for expected services with attached policies and tokens
+			require.Equal(t, len(c.expServices), len(obsServices))
+			require.ElementsMatch(t, c.expServices, obsServices)
+
+			// remove all the ECS tasks
+			lister.ECSClient = &mocks.ECSClient{}
+			resources, err = lister.List()
+			require.NoError(t, err)
+
+			// call Reconcile() to remove policies and tokens
+			for _, r := range resources {
+				s := r.(*ServiceInfo)
+				require.NoError(t, s.Reconcile())
+			}
+
+			// ensure cluster state matches previous
+			newPolicies, afterTokens, err := listACLs(consulClient, c.partition)
+			afterPolicies := make([]*api.ACLPolicyListEntry, 0, len(newPolicies))
+			for _, p := range newPolicies {
+				// we don't (currently) clean up namespaces nor their policies, so ignore them
+				if p.Name != "cross-namespace-read" && p.Name != "namespace-management" {
+					afterPolicies = append(afterPolicies, p)
+				}
+			}
+			require.NoError(t, err)
+			require.Equal(t, beforePolicies, afterPolicies)
+			require.Equal(t, beforeTokens, afterTokens)
+		})
+	}
+}
+
+// helper func that initializes a Consul test server and returns a Consul API client.
+func initConsul(t *testing.T) *api.Client {
```

Comment:
👍 I thought we had this, but it was something I added on my mesh-gateway branch (but slightly different for mesh-init).

https://github.com/hashicorp/consul-ecs/blob/286e28b2aaa6125c2eb7b970c9fd993f7f4aeaba/subcommand/mesh-init/command_test.go#L568-L584

I like these types of helpers that encapsulate a setup and a cleanup to declutter our test functions that tend to be quite long.